src/fhirtypes/common.ts

```diff
@@ -41,22 +41,44 @@ export function setPropertyOnInstance(
       const key =
         pathPart.primitive && i < pathParts.length - 1 ? `_${pathPart.base}` : pathPart.base;
       // If this part of the path indexes into an array, the index will be the last bracket
-      const index = getArrayIndex(pathPart);
+      let index = getArrayIndex(pathPart);
+      let sliceName: string;
       if (index != null) {
         // If the array doesn't exist, create it
         if (current[key] == null) current[key] = [];
+        sliceName = getSliceName(pathPart);
+        if (sliceName) {
+          const sliceIndices: number[] = [];
+          // Find the indices where slices are placed
+          current[key].forEach((el: any, i: number) => {
+            if (el?._sliceName === sliceName) {
+              sliceIndices.push(i);
+            }
+          });
+          // Convert the index in terms of the slice to the corresponding index in the overall array
+          if (index >= sliceIndices.length) {
+            index = index - sliceIndices.length + current[key].length;
+          } else {
+            index = sliceIndices[index];
+          }
+        }
         // If the index doesn't exist in the array, add it and lesser indices
         // Empty elements should be null, not undefined, according to https://www.hl7.org/fhir/json.html#primitive
         for (let j = 0; j <= index; j++) {
           if (j < current[key].length && j === index && current[key][index] == null) {
             current[key][index] = {};
           } else if (j >= current[key].length) {
-            current[key].push(j === index ? {} : null);
+            // _sliceName is used to later differentiate which slice an element represents
+            const arrayFiller = sliceName ? { _sliceName: sliceName } : null;
+            current[key].push(j === index ? {} : arrayFiller);
           }
         }
         // If it isn't the last element, move on, if it is, set the value
         if (i < pathParts.length - 1) {
           current = current[key][index];
+          if (sliceName) {
+            current._sliceName = sliceName;
```

Comment:
How does this approach work if the thing we're slicing is a primitive?  Does it?

Reply:
Actually, slicing primitives is real weird.  I'm not sure I care if it works or not.